map/local_ads_manager.cpp

```diff
@@ -274,6 +275,51 @@ void LocalAdsManager::UpdateViewport(ScreenBase const & screen)
   }
 }
 
+bool LocalAdsManager::DownloadCampaign(MwmSet::MwmId const & mwmId, std::vector<uint8_t> & bytes)
+{
+  bytes.clear();
+
+  std::string const url = MakeCampaignDownloadingURL(mwmId);
+  if (url.empty())
+    return true;
+
+  // Skip already downloaded campaigns. We do not lock whole method because RunHttpRequest
+  // is a synchronous method and may take a lot of time. The case in which countryName will
+  // be added to m_campaigns between locks is neglected.
+  {
+    std::lock_guard<std::mutex> lock(m_mutex);
+    auto const & countryName = mwmId.GetInfo()->GetCountryName();
+    if (m_campaigns.find(countryName) != m_campaigns.cend())
+      return false;
+  }
+
+  platform::HttpClient request(url);
+  bool const success = request.RunHttpRequest() && request.ErrorCode() == 200;
+
+  std::lock_guard<std::mutex> lock(m_mutex);
+  if (!success)
+  {
+    auto const it = m_failedDownloads.find(mwmId);
+    if (it == m_failedDownloads.cend())
+    {
+      m_failedDownloads.insert(std::make_pair(
+          mwmId, BackoffStats(Now(), kFailedDownloadingTimeout, 1 /* m_attemptsCount */)));
+    }
+    else
+    {
+      // Here we increase timeout as power 2 function.
+      it->second.m_currentTimeout = std::chrono::seconds(it->second.m_currentTimeout.count() << 1);
```

Comment:
+1 for @milchakov 
I like bithacks, but this is not the case. ```* 2``` is more readable. Let the compiler do the job.